keps/sig-network/3178-iptables-cleanup/README.md

```diff
@@ -0,0 +1,687 @@
+with `"KUBE-MARK-MASQ"`, to make the plugin use Kubernetes's iptables
+rules instead of creating its own. ([CRI-O] also assumes that
+`KUBE-MARK-MASQ` will exist, and uses it for implementing HostPort
+mappings.) Although in most cases kube-proxy will also create
+`KUBE-MARK-MASQ`, kube-proxy may not start up until after some other
+components, and some users may be running a network plugin that has
+its own proxy implementation rather than using kube-proxy. Thus, users
+may end up in a situation where some software is trying to use a
+`KUBE-MARK-MASQ` chain that does not exist.
+
+Most of these external components have simply copied kube-proxy's use
+of `KUBE-MARK-MASQ` without understanding why it is used that way in
+kube-proxy. But because they are generally doing much less IPTables
+processing than kube-proxy does, they could fairly easily be rewritten
+to not use the packet mark, and just have slightly-redundant
+`PREROUTING` and `POSTROUTING` rules instead.
+
+[CNI portmap plugin]: https://github.com/containernetworking/plugins/tree/master/plugins/meta/portmap
+[CRI-O]: https://github.com/cri-o/cri-o
+
+#### `iptables-wrapper`
+
+Another problem is the [iptables-wrapper script] for detecting whether
+the system is using iptables-legacy or iptables-nft. Currently it
+assumes that kubelet will always have created _some_ iptables chains
+before the wrapper script runs, and so it can decide which iptables
+backend to use based on which one kubelet used. If kubelet stops
+creating chains, then iptables-wrapper may find itself in a situation
+where there are no chains in either set of tables.
+
+```
+<<[UNRESOLVED iptables-wrapper-help ]>>
+
+It might make sense to have kubelet continue to create a single
+IPTables chain that exists for no purpose other than to indicate to
+pod-based external components whether the system is using
+iptables-legacy or iptables-nft.
+
+<<[/UNRESOLVED]>>
+```
+
+[iptables-wrapper script]: https://github.com/kubernetes-sigs/iptables-wrappers
+
+## Design Details
+
+### Implementation
+
+We cannot remove `KUBE-MARK-DROP` from kubelet until we know that
+kubelet cannot possibly be running against a version of kube-proxy
+that requires it.
+
+Thus, the process will be:
+
+#### Alpha
+
+We will document, in the appropriate places, the fact that external
+components should not be assuming the existence of Kubernetes's
+IPTables chains. (This probably includes a release note and a blog
+post.)
+
+Kube-proxy will be updated to not use `KUBE-MARK-DROP`, as described
+above. (This change is unconditional; it is _not_ feature-gated.) We
+should also ensure that kpng gets updated.
+
+Kubelet's behavior will not change by default, but if you enable the
+`IPTablesCleanup` feature gate, then:
+
+  1. It will stop creating `KUBE-MARK-DROP`, `KUBE-MARK-MASQ`,
+     `KUBE-FIREWALL`, and `KUBE-POSTROUTING`.
+
+  2. It will delete the `KUBE-MARK-DROP` and `KUBE-FIREWALL` chains,
```

Comment:
Do we need to actively delete them?  Or can we just rely on reboots to phase them out?